arch: arc: fix for hs eret has no copy of pc in interrupt entry

IRISZZW opened the pull request:
According to he Programmer's Reference Manual for ARC EM and HS,
there are some inconsistent performace between EM and HS in interrupt
entry. eret has copy of PC in interrupt entry in EM while not in HS,
so we should rely on eret register in interrupt entry, which will cause
carash for HS.

```
+----+---------------+---------------+--------------+
|    | FIRQ          | RIRQ          | RIRQ(Secure) |
+----+---------------+---------------+--------------+
| HS | ILINK=PC      | ILINK=PC      | NULL         |
+----+---------------+---------------+--------------+
| EM | ILINK=ERET=PC | ILINK=ERET=PC | ILINK=PC     |
+----+---------------+---------------+--------------+
```

similar fix: [arch: arc: for fast irq ERET has no copy of ilink](https://github.com/zephyrproject-rtos/zephyr/pull/18288/commits/7ccabd16b3aa6baa91b0d63a5224588e38802001)

Fixes: #29916

ruuddw reviewed (approved):
The wrapper is few cycles shorter/faster as well, now 👍 

Can you please also add a test/configuration (HS with FIRQ but 1 register bank) to ensure we catch problems like this in the future?

abrodkin reviewed (changes requested):
I'd re-worded a bit your commit message here, something along the lines of the following:
> According to the PRMs of both ARC EM & ARC HS families on entry to Fast IRQ handler ARC hardware saves PC (Program Counter) value of where processor was right before jumping to the IRQ handler into 2 registers: ILINK & ERET.
>
> But it turned out in case of ARC HS (at least in configuration with Fast IQs & 1 register bank) only ILINK was populated with the previous PC, while in Zephyr we relied on what we read out of ERET. That lead to funny issues when CPU returned from IRQ handling to some unexpected location.
>
> And now with that precious knowledge we're switching to return address recovery from ILINK so that with both families of ARC processors (EM & HS) we may get reliably good results.

Also it worth mention (so rightfully pointed to by @ruuddw) that due to the nature or `ILINK` (it's a core register as opposed to `ERET` being an AUX register) we may shave off another extra instruction for transferring `ERET` value from its AUX reg to a scratch core register to be later stored in the memory. And we do it in a fast path of a relatively often executed code (on every IRQ handling), which is really good.

IRISZZW committed:
arch: arc: fix for hs eret has no copy of pc in interrupt entry

According to the PRMs of both ARC EM & ARC HS families on entry
to Fast IRQ handler ARC hardware saves PC (Program Counter) value
of where processor was right before jumping to the IRQ handler into
2 registers: ILINK & ERET.

But it turned out in case of ARC HS (at least in configuration with
Fast IRQs & 1 register bank) only ILINK was populated with the
previous PC, while in Zephyr we relied on what we read out of ERET.
That lead to funny issues when CPU returned from IRQ handling
to some unexpected location.

And now with that precious knowledge we're switching to return
address recovery from ILINK so that with both families of ARC
processors (EM & HS) we may get reliably good results.

The wrapper is few cycles shorter/faster as well, as we may shave off
another extra instruction for transferring ERET value from its AUX reg
to a scratch core register to be later stored in the memory.

+----+---------------+---------------+--------------+
|    | FIRQ          | RIRQ          | RIRQ(Secure) |
+----+---------------+---------------+--------------+
| HS | ILINK=PC      | ILINK=PC      | NULL         |
+----+---------------+---------------+--------------+
| EM | ILINK=ERET=PC | ILINK=ERET=PC | ILINK=PC     |
+----+---------------+---------------+--------------+

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
